lib/logger_splunk_backend.ex

```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
+        {:ok, buffer_event(level, msg, ts, md, state)}
+
+      true ->
+        # we're in the middle of the request, but have reached our max buffer size
+        state = buffer_event(level, msg, ts, md, state)
+        {:ok, await_response(state)}
     end
-    {:ok, state}
   end
 
   def handle_event(:flush, state) do
-    # pretend the buffer is full to ensure all messages are sent
-    state = %{state | buffer_size: state.max_buffer}
-    {:ok, maybe_send(state)}
+    {:ok, await_response(state)}
+  end
+
+  @impl :gen_event
+  def handle_info(%{id: ref} = response, %{response: %{id: ref}} = state) do
+    {:ok, handle_http_response(response, state)}
   end
 
-  @impl true
-  def handle_info({:io_reply, _ref, :ok}, state) do
+  def handle_info({:EXIT, _pid, _reason}, state) do
     # ignored
     {:ok, state}
   end
+
   def handle_info(message, state) do
-    IO.puts("#{__MODULE__} unhandled message: #{inspect message}")
+    IO.puts(state.error_device, [
+      "ERROR ",
+      inspect(__MODULE__),
+      " unhandled message: ",
+      inspect(message)
+    ])
+
     {:ok, state}
   end
 
+  defp handle_http_response(%HTTPoison.AsyncStatus{code: 200}, state) do
+    state
+  end
+
+  defp handle_http_response(%HTTPoison.AsyncStatus{code: invalid}, state) do
+    IO.puts(state.error_device, [
+      "ERROR unexpected status code when logging to Splunk: ",
+      inspect(invalid),
+      "\n",
+      "ERROR sent: ",
+      state.output
+    ])
+
+    %{state | printing_response?: true}
+  end
+
+  defp handle_http_response(%HTTPoison.AsyncHeaders{}, state) do
+    state
+  end
+
+  defp handle_http_response(%HTTPoison.AsyncChunk{}, %{printing_response?: false} = state) do
+    state
+  end
+
+  defp handle_http_response(
+         %HTTPoison.AsyncChunk{chunk: body},
+         %{printing_response?: true} = state
+       ) do
+    IO.puts(state.error_device, ["ERROR response: ", body])
+    state
+  end
+
+  defp handle_http_response(%HTTPoison.AsyncEnd{}, state) do
+    maybe_send(%{state | response: nil, output: nil, printing_response?: false})
+  end
+
+  defp handle_http_response(%HTTPoison.Error{reason: reason}, state) do
+    IO.puts(state.error_device, [
+      "ERROR unable to log to Splunk: ",
+      inspect(reason),
+      "\n",
+      "ERROR sent: ",
+      state.output
+    ])
+
+    maybe_send(%{state | response: nil, output: nil, printing_response?: false})
+  end
+
+  defp meet_level?(_level, nil) do
+    true
+  end
+
+  defp meet_level?(level, min_level) do
+    Logger.compare_levels(level, min_level) != :lt
+  end
+
   defp log_event(level, msg, ts, md, state) do
+    state = buffer_event(level, msg, ts, md, state)
+    maybe_send(state)
+  end
+
+  defp buffer_event(level, msg, ts, md, state) do
     data = format_message(msg, level, ts, md, state)
-    state = %{
-      state |
-      buffer: [data | state.buffer],
-      buffer_size: state.buffer_size + 1
+
+    %{
+      state
+      | buffer: [state.buffer, data],
```

Comment:
Not sure it's worth the optimization, but would it be faster if we did `buffer: [data | state.buffer]` here instead, and then did an `Enum.reverse` when it came time to actually send it? For that matter, since the event actually has a timestamp in it anyway, which Splunk will use to order properly, maybe we wouldn't even need the reversal.

Reply:
This creates an IO list (https://www.bignerdranch.com/blog/elixir-and-io-lists-part-1-building-output-efficiently/) which then gets converted to a binary all at once here: https://github.com/mbta/logger_splunk_backend/pull/6/files#diff-422fe5b31c1b6f41cc85cb125846e02cR229

For example:
```elixir
iex(1)> iolist = [[[["1" | "2"], "3"], "4", ["5", "6"]], "7", "8"]
[[[["1" | "2"], "3"], "4", ["5", "6"]], "7", "8"]
iex(2)> IO.iodata_to_binary(iolist)
"12345678"
```